### ISSUE TITLE
Object IDs created by BSON v4 are silently ignored by Node driver 3.x

### ISSUE THREAD

**Issue by justingrant** [mongodb/js-bson#306]
The latest Node driver 3.1.3 ignores all `ObjectId` instances created by the v4 BSON library, including Object IDs put in filters and Object IDs used as values to be stored.  The Object ID instance(s) will be silently ignored by the Node driver and will never be sent to the server.  No errors are generated and the server operation will complete successfully as if the `ObjectId` was not included in the input. 

This bug essentially makes the 4.x library incompatible with MongoDB, because any non-trivial MongoDB app will involve sending `ObjectId` instances to the server.  Using `ObjectId` inside a filter document is particularly dangerous. For example I found a case where, in a `findOneAndUpdate`, a 4.x `ObjectId` was in a filter for a `$pull` operation.  Instead of removing only one item from the array, MongoDB removed all of them!

Repro is trivial: 
* Create a document containing an BSON 4.x `ObjectId` instance, e.g. `{a: 1, b: new ObjectId()}`
* Try to insert that document using the Node 3.1.3 driver or use it as a query filter
* Expected: entire document is inserted / entire document is used a filter
* Actual: 
  * (insert case) only `{a: 1}` is inserted. `b` is not inserted, but the insert succeeds and no errors are returned. 
  * (filter case) all documents matching `{a: 1}` are included in the filter. The `ObjectId` filter is ignored.  This can lead to unexpected data loss for filters on delete or update operations.

The root cause is the switch of `_bsontype` from `ObjectID` to `ObjectId` (lowercase "d") in BSON v4.  When the Node driver 3.1.3 serializes commands before sending them to the server, the BSON v1 serializer used by Node driver 3.1.3 doesn't recognize `ObjectId` instances created by BSON V4. The serializer lacks a default "fall-through" case so`ObjectId` instances (and any other objects with an unrecognized `_bsontype`) are silently ignored by the serializer without any errors.  From the client's perspective, this just looks like data loss.

I can see a few possible fixes for this: 
* Option 1: go back to using `{_bsontype: 'ObjectID'}` like BSON v1 used.  Unless there's a compelling reason to leave it as `ObjectId`, this seems like the cleanest solution because it allows interop between BSON v4 and current Node drivers (or any other library that creates Object IDs that are duck-type compatible with BSON v1 ObjectID instances).
* Option 2: Patch the BSON 1.x branch to recognize `_bsontype: 'ObjectId'` too, and then patch the 3.x node driver to depend on the patched BSON 1.x. This is essentially the reverse of #290 which added `ObjectID` support to BSON 4.x.
* Option 3: Patch the 3.x Node driver to depend on BSON 4.x, and document the requirement for users to permanently avoid mixing BSON v4 with older versions of the Node driver that use BSON 1.x.  I assume that this would be the riskiest option given that there may be other potential bugs or breaking changes that are lurking in BSON v4 that haven't been found yet, so it might be too risky to bump this dependency without a minor or major release of the Node driver. 

@mbroadst - I'm happy to help with PRs for any and/or all of these options but could use your guidance. Do you have a preference for which option(s) should be implemented?  Option 1 seems like the least risky and impactful because there are few users of BSON v4 but many of BSON v1, but I don't understand the context behind the shift to `_bsontype: 'ObjectId'` so I may be wrong. Why was `_bsontype` changed in the first place?  I understand the logic of changing the class name to `ObjectId`, but why did `_bsontype` have to change too?  Is there any harm in changing it back?

The code below (from the BSON 1.x branch) is where the problem happens. If an object looks like `{_bsontype: 'ObjectId', ... }`, then execution falls through to the end of the chain of `if`/`else` statements without being serialized but also without generating any errors.  (BTW, the lack of a fall-through case for _bsontype is also a bug in 4.x. I'll file a new issue & PR to address this problem in 4.x.)

https://github.com/mongodb/js-bson/blob/39215f038b6787d015bdd1525e4b9db060603350/lib/bson/parser/serializer.js#L919-L982

**Comment by justingrant** [comment]
@mbroadst @daprahamian - got any opinions for how you'd like to see this bug fixed?  I have time later today that I can use to build a PR for this issue, but need guidance before proceeding.

BTW, turns out the problem is worse than I originally thought-- Object IDs used as filters are also ignored, which can cause unexpected data loss if the filter was on a delete or update operation.

**Comment by mbroadst** [comment]
Hey @justingrant, we usually have our team meeting on Thursdays so I deferred discussion until today about it. Option 2 seems to be the universally preferred option here, as it doesn't actively break any user code. Thanks for pointing this issue out, I made [NODE-1873](https://jira.mongodb.org/browse/NODE-1873) to track this issue

**Comment by justingrant** [comment]
No worries, thanks @mbroadst for the update. I'll send an "Option 2" PR on the 1.x branch soon-- probably tomorrow.

In parallel with Option 2, does it make sense to *also* do Option 1 (switch back to `_bsonType: 'ObjectID'`) as well?  Even once both 4.x and 1.x BSON libraries handle both variants, GitHub search shows ~3K [public code files on GitHub](https://github.com/search?q=_bsontype+ObjectID&type=Code) that currently use or depend on `_bsonType` being `'ObjectID'`. Surely there are many more in private repos. At least some of this code is going to be broken once the 4.x library starts getting more adoption.

Seems like a safer choice for the ecosystem would be to simply leave `_bsontype` unchanged from what it was in 1.x.  Is there another reason why it's important to have `_bsonType` === `'ObjectID'` in the 4.x library?

**Comment by mbroadst** [comment]
@justingrant I was trying to prevent leaving 4.0.0 users in the lurch, but on further discussion I think we can go ahead and make the proposed changes in option 1. Would you like to put that PR together, or shall I just merge it?

**Comment by justingrant** [comment]
Cool. I can do the PR. Below is my current plan for what goes in the PR. Holler if you see a problem: 
* revert `_bsontype` (option 1 above) in the `ObjectId` class
* re: "leaving 4.00 users in the lurch" I have an idea how to support 4.0.0 users too with very little extra code. I'll include it unless tests expose problems.
* add tests for all three ObjectID cases: the current class, the 4.0.0 class, and the class currently used by `require('mongodb')`
* add tests for all 1.x types (not just ObjectID, also Binary, Long, etc.) to make sure that similar cross-version compatibility issues don't show up in those types. Fix any problems that these tests might uncover.
